### DISCUSSION TITLE
In-memory cache for DataSourceInstanceSettings

### DISCUSSION THREAD

**Discussion by joanlopez** [grafana/grafana#50200] [Platform]
## Context

_(This is not directly related with recent envelope encryption updates, but with encryption in general)_

During the investigation of https://ops.grafana.net/a/grafana-incident-app/incidents/129, where query latency went out of SLO, @csmarchbanks identified **a ton of CPU usage is spent in decryption** ([conprof link](https://admin-prod-us-central-0.grafana.net/conprof/hosted-grafana/pprof/e19fbmFtZV9fPSJwcm9maWxlIixjb250YWluZXI9ImdyYWZhbmEiLGluc3RhbmNlPSIxMC4xNDQuMjEzLjcyOjYwNjAiLGpvYj0iY29ucHJvZiIsbmFtZXNwYWNlPSJob3N0ZWQtZ3JhZmFuYSIscG9kPSJwaXNtby1ncmFmYW5hLTU2OWM2NmJmNDctd3FtMngiLHNsdWc9InBpc21vIn0=/1654272279808/flamegraph)):

<img src="https://user-images.githubusercontent.com/5459617/171964453-428d79e9-67d4-4fd1-b9a3-47171d207dfe.png" width=40% height=40%>

By analyzing the details, seems the decryption is **happening within a call to [`adapters.ModelToInstanceSettings`](https://github.com/grafana/grafana/blob/main/pkg/plugins/adapters/adapters.go)** which is a function that transforms the `models.DataSource` into `backend.DataSourceInstanceSettings`, which is the data structure used to send data source settings (name, url, secure json data, etc) to backend plugins.

So, **(i) it seems that function is called quite often** (some callers are: check data source health, node execution, handle query data, some data sources getters), and **(ii) decryption happens on every call to that function** because although the data source is usually obtained from an in-memory cache, secure json data present in data source model is not decrypted before storing it into the cache, therefore the decryption needs to happen at transformation time.

## Proposals

See below **two different proposals based on in-memory caching**, as suggested by @hairyhenderson and @csmarchbanks during the incident investigation, and also as it's used for other use cases. 

A benefit (pro) from both is that it would help on reducing the amount of decryption operations, which seems to be an (intentionally) CPU intensive operation.

### Proposal 1: Cache data sources with secure json data decrypted

As a cache for data sources is already being used, **we could add a decryption step before adding new entries into that,** so data sources would be cached with the secure json data already decrypted.

**Pros**
Simpler to implement because it won't require new cache, decryption would be kept together with data sources fetching code and it would be transparent for `adapters.ModelToInstanceSettings` as well as for other functions that may potentially be experiencing similar behavior (getting the data source from cache but decrypting on every call).

**Cons**
As stated above (and detailed in incident's Slack thread), decryption operations are CPU intensive (intentionally for brute-force safety), so the risk of this strategy would be that it may represent a considerable added overhead for many data source operations that don't actually need the secure json data to be decrypted (like a simple `GET /datasource/:id`, among others).

### Proposal 2: Add a new cache specifically for DataSourceInstanceSettings 

Similarly to what we're already doing with data sources, we could add a new in-memory cache with data sources already transformed into `DataSourceInstanceSettings`, making it possible to cache the results of decryption as well as json marshaling.

This would be a service similar to `datasources.CacheService` (not sure where would be defined) that probably would be injected into different places (potentially to all the `adapters.ModelToInstanceSettings`) callers.

**Pros**
As this would be a specific solution for the use case where `models.DataSource` structs are transformed into `backend.DataSourceInstanceSettings`, we would be able to only add the overhead of decrypting secure json data when actually needed, in contrast to the previous proposal.

**Cons**
In contrast to the previous proposal, this would be harder to implement because it'd imply creating a new cache service with bit of extra logic, injecting it into different places and adjusting the calls to `adapters.ModelToInstanceSettings`. 
Also, as `adapters.ModelToInstanceSettings` is currently being called as a global function, there's risk that it's not that easy to inject the new cache service in certain places where that function is currently being used.
Finally, another con (although maybe it's negligible), is that it would mean partially duplicated data in memory (because `backend.DataSourceInstanceSettings` is basically a transformation of `models.DataSource`, which is already being cached).

Any kind of input would be appreciated. Thanks! 🙌🏻

**Comment by joanlopez** [comment]
cc/ @csmarchbanks @hairyhenderson @undef1nd @marefr @yoziru for visibility, because you've been directly or indirectly involved in the incident investigation, as well as @sakjur @papagian @marefr because I think you're code owners and/or familiar with it (feel free to correct me if wrong and/or to suggest any other potentially interested peer).

Thanks so much! 🙏🏻

**Reply by marefr** [reply]
Pinging @guicaulada @mmandrus since they've recently introduced the secrets store and refactored quite a lot of things related to this. We had some discussions regarding caching of decrypted secrets since we've had that in place since https://github.com/grafana/grafana/pull/20579. They moved the decryptioncache from datasouce_service.go to the secrets store implementation I think (Ref https://github.com/grafana/grafana/pull/45804/files#diff-e75b390a62e372c0e3f8956d05dd9515acdd67d6fce46b4984afdf453e473adeL255).

Regarding the proposals, 1) I guess is what we already had in place via https://github.com/grafana/grafana/pull/20579? Quite hard to follow with all the changes that has happened to encryption/secrets in general over the last couple months so those that been involved need to figure things out I think. 

> so the risk of this strategy would be that it may represent a considerable added overhead for many data source operations that don't actually need the secure json data to be decrypted (like a simple GET /datasource/:id, among others).

This one is important really. Decryption needs to be lazy, e.g. to only happen when needed.

**Comment by jacksongoode** [comment]
Just out of curiosity, I had a similar discussion post https://github.com/grafana/grafana/discussions/47836 and was wondering if a decrypted cache would allow for editing of the json model directly? I might be far off with this. Being able to edit the dashboard's json model directly and see live changes would be a great consequence of this if so.

**Reply by joanlopez** [reply]
Hey @jacksongoode, 
After taking a quick look to the discussion you mentioned, it seems to be completely different kind of cache (browser vs server).
So, I'm not sure both discussions are actually related.
Anyway, thanks for opening that discussion, I think there's been some progress in that direction (local "draft" changes).

**Reply by jacksongoode** [reply]
@joanlopez Thank you for the response! If there are other discussions or PR's related to the topic I would love to check them out and comment.

**Comment by papagian** [comment]
Is the decryption cost similar/comparable for the OSS and the different secret engines? The above profile which kind of decryption is using?

**Reply by joanlopez** [reply]
> Is the decryption cost similar/comparable for the OSS and the different secret engines? The above profile which kind of decryption is using?

There are multiple aspects to consider here:

- If an external key management service (KMS) is configured for envelope encryption (Enterprise-only feature, NOT configured by default), certain encrypt/decrypt operations may require an additional network request to the external KMS. That operation is cached in-memory with a configurable TTL.
 
- If the secret is stored into an external secret store (Enterprise-only feature?, NOT configured by default; cc/ @guicaulada for confirmation), certain encrypt/decrypt operations may require an additional network request to the external secret store to fetch it. I guess that operation is cached in-memory as well.

- Finally, and after considering the above aspects (secret retrieval and data key decryption), the cost of all decryption operations should be similar/comparable, yes. It'll depend mostly on the length of the secret, I guess.

cc/ @undef1nd

**Reply by guicaulada** [reply]
> If the secret is stored into an external secret store (Enterprise-only feature?, NOT configured by default; cc/ @guicaulada for confirmation), certain encrypt/decrypt operations may require an additional network request to the external secret store to fetch it. I guess that operation is cached in-memory as well.

If there's a secret store plugin installed, all encryption, decryption and caching is handled by the plugin, in the case of AWS Secrets Manager for example, secrets are stored in plain text and retrieved as plain text, all encryption, decryption and caching is handled by AWS.

**Reply by joanlopez** [reply]
Hmmm.. that's interesting, could that somehow conflict with our own internal encryption? 🤔 I'm a bit afraid we might be trying to encrypt/decrypt certain data that we suppose is encrypted/decrypted while that should (only) be managed by external secrets manager in that case, I guess.

cc/ @undef1nd

**Reply by guicaulada** [reply]
@joanlopez that shouldn't be the case, the Unified Secrets skips the default encryption entirely when delegating to the plugin, the SecretStore.Get function only sends the request to the plugin via gRPC and awaits a response with the secret already decrypted.

At the moment Unified Secrets is only integrated to Data Sources, so nothing else should be impacted by this and should use the default encryption/decryption and storage method.

**Reply by joanlopez** [reply]
Gotcha! Thanks for clarifications! 💟 
Yeah, I agree that since we're only using it for data sources credentials, it should be safe and under control.

**Comment by guicaulada** [comment]
The decryption cache is implemented for secret retrieval in the code block starting on [line 59 of pkg/services/secrets/kvstore/sql.go](https://github.com/grafana/grafana/pull/45804/files#diff-47cf0d07bd838b569ac12071caf5f7325cbe56b2805d0573b59a239b72e0269eR59) and should prevent secrets from being decrypted unecessarily when the decryptedValues function is called on the decryptFn, maybe there's an issue with that functionality causing the decryption overhead?

**Reply by joanlopez** [reply]
> The decryption cache is implemented for secret retrieval in the code block starting on [line 59 of pkg/services/secrets/kvstore/sql.go](https://github.com/grafana/grafana/pull/45804/files#diff-47cf0d07bd838b569ac12071caf5f7325cbe56b2805d0573b59a239b72e0269eR59) and should prevent secrets from being decrypted unecessarily when the decryptedValues function is called on the decryptFn, maybe there's an issue with that functionality causing the decryption overhead?

I should revisit the code involved in both use cases but I think they're separate, more than an actual issue.

**Reply by marefr** [reply]
I did discuss this with @guicaulada last week and pointed out that [this](https://github.com/grafana/grafana/blob/624a3240f0d5fe3f762b196ea8d876fde77a1ab2/pkg/services/secrets/kvstore/sql.go#L44) will hit the database on each call even though it's cached which could overload a database in the worst of scenarios. We discussed introducing  a short-lived cache of some seconds, similar to how the [datasource cache](https://github.com/grafana/grafana/blob/624a3240f0d5fe3f762b196ea8d876fde77a1ab2/pkg/services/datasources/service/cache_service.go#L94) is implemented as a way to handle a large batch of datasource queries happening within a short period of time.

**Reply by guicaulada** [reply]
We have created a task to implement the short-lived cache and it's on our To Do list, it should be tackled soon, we are currently fixing some issues and improving migration and integration with secret store plugins.

**Reply by guicaulada** [reply]
I started working on the short-lived cache for the unified secrets.

**Reply by guicaulada** [reply]
I created the PR https://github.com/grafana/grafana/pull/51275 with the short-lived cache implementation, let me know what you think!

**Comment by hairyhenderson** [comment]
FYI this is affecting us again, in a different instance... Almost exactly the same flame graph 😓

**Reply by marefr** [reply]
What Grafana version?

**Reply by hairyhenderson** [reply]
This was with v9.0.0

**Reply by marefr** [reply]
Just merged https://github.com/grafana/grafana/pull/54146 targeted for 9.1.2 and should hopefully resolve this.

**Reply by joanlopez** [reply]
Excellent! Thanks @marefr for taking care of it! 😊

**Reply by hairyhenderson** [reply]
Thanks so much @marefr, I appreciate it! 🎉

**Comment by marefr** [comment]
I've looked a bit more on the flame graph and seems like this is due to below of code introduced in #44862:
https://github.com/grafana/grafana/blob/370d6a6f7b23ab4ff867a9455b590e4ea4174871/pkg/services/ngalert/eval/eval.go#L200

I see two problems with this code:
1. Seems like it was introduced to resolve a bug, but to me I think the problems should have been fixed in the datasources instead.
2. It uses the secrets service `secrets.Service` which feels wrong based on above discussions. Feels like this perhaps should use the `kvstore.SecretsKVStore` to make use of cached decrypted data since it now seems to go directly through the envelope encryption service without cache?

Fixing 1) is probably the most sane solution, but not sure if it's a risk introducing same problem that it originally solved. Regarding 2) feels like there's currently hard for developers to understand when to make use of `kvstore.SecretsKVStore` or  `secrets.Service`, unless I misunderstand something. Would it be possible to refactor so there's only one secrets service to rule them all that are used everywhere?

@guicaulada @joanlopez WDYT?

**Reply by guicaulada** [reply]
I agree fixing 1 is the most sane solution. Regardin 2, there's a plan to eventually integrate every service with the SecretsKVStore and eventually deprecating the old secrets service, however it was too big of a refactor to do at once, we are trying to finish the plugin part of unified secrets to deliver on other priorities so we can finally come back and finish integrating unified secrets on other services.

**Reply by marefr** [reply]
Opened https://github.com/grafana/grafana/pull/51237 to get the discussion/work going to resolve this

**Reply by marefr** [reply]
Closed that one in favor of https://github.com/grafana/grafana/pull/51061 for now.

**Reply by marefr** [reply]
@aocenas given https://github.com/grafana/grafana/pull/51061 is closed and there's been additional Prometheus refactor happening feels like we should be able to remove the code referenced above that adds custom headers. WDYT?

**Reply by marefr** [reply]
Opened https://github.com/grafana/grafana/pull/54146 to resolve this

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This seems to be resolved, so I am going to close this discussion.
